tools/tsp-client/src/network.ts

```diff
@@ -0,0 +1,92 @@
+import { createDefaultHttpClient, createPipelineRequest } from "@azure/core-rest-pipeline";
+import { createFileTree } from "./fileTree.js";
+import { resolveImports } from "./typespec.js";
+import { Logger } from "./log.js";
+
+const httpClient = createDefaultHttpClient();
+
+async function fetch(url: string, method: "GET" | "HEAD" = "GET"): Promise<string> {
+  const result = await httpClient.sendRequest(createPipelineRequest({ url, method }));
+  if (result.status !== 200) {
+    throw new Error(`failed to fetch ${url}: ${result.status}`);
+  }
+  return String(result.bodyAsText);
+}
+
+export function isValidUrl(url: string) {
+  try {
+    new URL(url);
+    return true;
+  } catch {
+    return false;
+  }
+}
+
+export function doesFileExist(url: string): Promise<boolean> {
+  return fetch(url, "HEAD")
+    .then(() => true)
+    .catch(() => false);
+}
+
+export function rewriteGitHubUrl(url: string): string {
+  if (url.includes("github.com")) {
+    const result = url.replace("github.com", "raw.githubusercontent.com");
+    Logger.debug(`rewriting github url to direct link: ${result}`);
+    return result;
+  }
+  return url;
+}
+
+export async function downloadTsp(rootUrl: string) {
+  const seenFiles = new Set<string>();
+  const moduleImports = new Set<string>();
+  // fetch the root file
+  const filesToProcess = [rootUrl];
+  seenFiles.add(rootUrl);
+  const fileTree = createFileTree(rootUrl);
+
+  while (filesToProcess.length > 0) {
+    const url = filesToProcess.shift()!;
+    const sourceFile = await fetch(url);
```

Comment:
Depending on the number of files fetching them each individually from github has proven to be error prone due to rate limiting. That is why the PS version of the script did a partial git clone instead.

Reply:
I was talking to Laurent yesterday and we'll start off with the sparse clone for now and mimic what the pwsh scripts are doing

Reply:
Yeah, not too difficult to support both ways of doing it, this was just the fastest way I could figured out which worked for any web-accessible URL rather than something that needs a git repo

Reply:
If examples of sparse cloning would be helpful you can find some here: https://github.com/Azure/azure-sdk-tools/blob/main/eng/common/pipelines/templates/steps/sparse-checkout.yml#L46-L75, however you likely won't need to set `--no-cone` mode.